modules/auresamp/auresamp.c

```diff
@@ -0,0 +1,234 @@
+ *
+ * Copyright (C) 2022 Commend.com - c.spielberger@commend.com
+ */
+
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+
+enum {
+	MAX_PTIME       =    60,  /* Maximum packet time in [ms] */
+};
+
+/**
+ *  The auresamp module is one of the audio filters. The order of the filters
+ *  is specified by the order in the config file.
+ *
+ *  .    .-------.   .----------.   .-------.   .---------.
+ *  |    |       |   |          |   |       |   |         |
+ *  |O-->| ausrc |-->| auresamp |-->| aubuf |-->| encoder |--> RTP
+ *  |    |       |   |          |   |       |   |         |
+ *  '    '-------'   '----------'   '-------'   '---------'
+ *
+ *       .--------.   .-------.   .----------.   .--------.
+ * |\    |        |   |       |   |          |   |        |
+ * | |<--| auplay |<--| aubuf |<--| auresamp |<--| decode |<-- RTP
+ * |/    |        |   |       |   |          |   |        |
+ *       '--------'   '-------'   '----------'   '--------'
+ */
+
+struct auresamp_st {
+	union {
+		struct aufilt_enc_st eaf;
+		struct aufilt_dec_st daf;
+	};                       /* inheritance              */
+
+	int16_t *sampv;          /* s16le audio data buffer  */
+	int16_t *rsampv;         /* resampled data           */
+	struct auresamp resamp;  /* resampler                */
+	struct aufilt_prm oprm;  /* filter output parameters */
+};
+
+
+static void destructor(void *arg)
+{
+	struct auresamp_st *st = arg;
+
+	mem_deref(st->rsampv);
+	mem_deref(st->sampv);
+}
+
+
+static int sampv_alloc(struct auresamp_st *st, struct auframe *af)
+{
+	size_t psize;
+
+	psize = MAX_PTIME * af->srate * af->ch * sizeof(int16_t) / 1000;
+	st->sampv = mem_zalloc(psize, NULL);
+
+	if (!st->sampv)
+		return ENOMEM;
+
+	return 0;
+}
+
+
+static int resamp_setup(struct auresamp_st *st, struct auframe *af)
+{
+	int err = 0;
+	size_t psize;
+
+	err = auresamp_setup(&st->resamp, af->srate, af->ch,
+			     st->oprm.srate, st->oprm.ch);
+	if (err) {
+		warning("resample: auresamp_setup error (%m)\n", err);
+		return err;
+	}
+
+	psize = MAX_PTIME * st->oprm.srate * st->oprm.ch * 2 / 1000;
+
+	st->rsampv = mem_deref(st->rsampv);
+	st->rsampv = mem_zalloc(psize, NULL);
+	if (!st->rsampv)
+		return ENOMEM;
+
+	return 0;
+}
+
+
+static int common_update(struct auresamp_st **stp, struct aufilt_prm *oprm)
+{
+	struct auresamp_st *st;
+	if (!stp || !oprm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	st = mem_zalloc(sizeof(*st), destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->oprm = *oprm;
+	auresamp_init(&st->resamp);
+
+	*stp = st;
+	return 0;
+}
+
+
+static int common_resample(struct auresamp_st *st, struct auframe *af)
+{
+	size_t outc;
+	int16_t *sampv = af->sampv;
+	int err = 0;
+
+	if (st->oprm.srate == af->srate && st->oprm.ch == af->ch)
+		return 0;
+
+	if (af->fmt != AUFMT_S16LE) {
+		if (!st->sampv)
+			err = sampv_alloc(st, af);
+
+		if (err)
+			return err;
+
+		auconv_to_s16(st->sampv, af->fmt, af->sampv, af->sampc);
+		sampv = st->sampv;
+	}
```

Comment:
I guess not. Because module auresamp will use `auresamp` of rem, like audio.c does so far. And this resampler only supports s16le. The audio conversion that we need here is from input format to s16le and after resampling to the output format.

Module `auconv` converts the samples to the configured `config_audio.play_fmt` and `config_audio.enc_fmt`. BTW I think we should also change the initialization of `auconv` module to a similar behavior like we plan for auresamp:
- use `struct auframe` for the input format
- use `struct aufilt_prm` for the output format

But the conversion to/from s16le in `auresamp` is only internally necessary.